File: go/p018.go
package main

import (
	"strconv"
	"strings"
	"fmt"
	"bufio"
	"os"
)

func strtoint(line string) []int {
	cells := strings.Split(line, " ")
	result := make([]int, len(cells))
	for i, v := range cells {
		result[i], _ = strconv.Atoi(v)
	}
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	f, _ := os.Open("../sources/p018.txt")
	scanner := bufio.NewScanner(f)

	puff := make([][]int, 0, 15)
	for scanner.Scan() {
		puff = append(puff, strtoint(scanner.Text()))
	}

	for i := len(puff) - 2; i >= 0; i-- {
		for j := 0; j < len(puff[i]); j++ {
			puff[i][j] += max(puff[i+1][j], puff[i+1][j+1])
		}
	}
	fmt.Println(puff[0][0])
	// output: 1074
}

File: python/p018.py
# Maximum path sum I
#
# find the maximum total from top to bottom of the triangle below:

with open("../sources/p018.txt") as f:
    source = f.read().strip()

grid = [[int(x) for x in row.split()] for row in source.splitlines()]

def max_path(y, x):
    if y < len(grid) - 1:
        return grid[y][x] + max(max_path(y+1,x), max_path(y+1,x+1))
    else:
        return grid[y][x]

print(max_path(0,0))

# output: 1074
